content/blog/rails-restapi-outside-in-tdd.md

```diff
@@ -0,0 +1,740 @@
+
+```ruby
+Failing Test Case:
+        it "invokes the instance of Create_Runner successfully and renders the json" do
+            params = {title: "Todo-1", description:'First todo' }
+            mock = double("TodosRunner")
+            expect(mock).to receive(:create_todo).with(params)
+            create_new_todo = TodosController.new(mock)
+            allow(create_new_todo).to receive(:todo_params).and_return(params)
+            allow(create_new_todo).to receive(:json_result).and_return(true)
+            create_new_todo.create
+        end
+
+$rspec
+
+RED:
+ 1) Todos Controller invokes the instance of Create_Runner successfully and renders the json
+     Failure/Error: allow(create_new_todo).to receive(:json_result).and_return(true)
+       #<TodosController:0x000000000069f0> does not implement: json_result
+GREEN:
+
+    def create
+        result, status = @runner.create_todo(todo_params)
+        json_result(result, status)
+    end
+    def json_result(object, status)
+        render json: object, status:status
+    end
+
+$ rspec
+
+All Greens
+
+```
+
+I went on to write the request spec(todos_spec.rb) that expects 422 when the title, which is a mandatory field, is not passed.
+
+```ruby
+Failing Test Case:
+  context 'with invalid request attributes ' do
+    it 'returns status code 422' do
+      post '/qwerts', params: {title: "", description:'First todo' }, as: :json
+      expect(response).to have_http_status(422)
+    end
+  end
+
+$ rspec
+
+RED:
+  1) Todos with invalid request attributes returns status code 422
+     Failure/Error: expect(response).to have_http_status(422)
+       expected the response to have status code 422 but it was 201
+```
+
+```ruby
+The fix has to be done at Runner layer as this layer contains the application logic.
+todos_runner_spec.rb
+Failing Test Case:
+        it " returns unprocessable entity when title is not passed" do
+            save_params = {title: "", description:'First todo' }
+            todo = Todo.new(save_params)
+            todosRunner = TodosRunner.new
+            todo, status = todosRunner.create_todo(save_params)
+            expect(status).to eq(:unprocessable_entity)
+            expect(todo["message"]).to eq("Invalid input values")
+        end
+Fix:
+    def create_todo(add_params)
+        todo =@repo.new_todo(add_params)
+        if(@repo.save_todo(todo))
+            return todo, :created
+        else
+            error = Hash.new
+            error["message"] = " Invalid input values"
+            return error ,:unprocessable_entity
+        end
+    end
+ $ rspec
+
+All Green
+
+```
+
+### It is always quite a sight to see all green
+
+I was so immersed into conversing with rspec to develop this code, I felt like I was a part of a play, playing a character and that the play is being observed, and I am responding only when there is a cue and refrained to when there is no cue as I may mislead them observers if I take any decision without the cue from the rspec.
```

Comment:
I was so immersed in conversing with rspec while developing this application that I kind of got distanced from reality. I felt like I’d embodied this character in a play that was being watched and I was to only act out my lines on cue lest I’d mislead the audience. Rspec called the shots here!